fw/http2.c

```diff
@@ -192,26 +192,147 @@ tfw_h2_apply_new_settings(TfwH2Ctx *ctx)
 	clear_bit(HTTP2_SETTINGS_NEED_TO_APPLY, ctx->settings_to_apply);
 }
 
-int
-tfw_h2_init(void)
+#define TFW_H2_PAGE_ORDER 2
+#define TFW_H2_BOUNDARY(base) 					\
+	(char *)base + PAGE_SIZE * (1 << TFW_H2_PAGE_ORDER);
+#define TFW_H2_NEXT_BLOCK(boundary)				\
+	(unsigned long *)(boundary - sizeof(unsigned long));
+
+TfwH2Ctx *
+tfw_h2_context_alloc(void)
+{
+	struct page *pg;
+
+	pg = alloc_pages(GFP_ATOMIC, TFW_H2_PAGE_ORDER);
+	if (!pg)
+		return NULL;
+	return (TfwH2Ctx *)page_address(pg);
+}
+
+void
+tfw_h2_context_free(TfwH2Ctx *ctx)
+{
+	free_pages((unsigned long)ctx, TFW_H2_PAGE_ORDER);
+}
+
+static inline void
+tfw_h2_context_init_stream_storage_impl(TfwH2Ctx *ctx, void *base,
+					unsigned long *next_block)
+{
+	TfwStream *stream = (TfwStream *)base;
+
+	while ((char *)stream <= (char *)next_block - sizeof(TfwStream)) {
```

Comment:
It seems here is memory corruption because of `<=`, in case when quantity is a multiple of size, the store will be filled without hole in the end that need for storing pointer to next block.

Reply:
Why if stream == next - sizeof(TfwStream) we have one place for struct stream. I am not right?